File: questModule/src/main/java/ru/nekit/android/qls/lockScreen/receiver/PhoneCallReceiver.java
package ru.nekit.android.qls.lockScreen.receiver;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.telephony.TelephonyManager;

import java.util.Date;

public class PhoneCallReceiver extends BroadcastReceiver {

    public static final String INCOMING_CALL_RECEIVED = "incoming_call_received";
    public static final String INCOMING_CALL_ENDED = "incoming_call_ended";
    public static final String OUTGOING_CALL_ENDED = "outgoing_call_ended";
    public static final String OUTGOING_CALL_STARTED = "outgoing_call_started";

    private static int lastState = TelephonyManager.CALL_STATE_IDLE;
    private static Date callStartTime;
    private static boolean isIncoming;
    private static String savedNumber;

    @Override
    public void onReceive(Context context, Intent intent) {
        if ("android.intent.action.NEW_OUTGOING_CALL".equals(intent.getAction())) {
            savedNumber = intent.getExtras().getString("android.intent.extra.PHONE_NUMBER");
        } else {
            String stateStr = intent.getExtras().getString(TelephonyManager.EXTRA_STATE);
            String number = intent.getExtras().getString(TelephonyManager.EXTRA_INCOMING_NUMBER);
            int state = 0;
            if (TelephonyManager.EXTRA_STATE_IDLE.equals(stateStr)) {
                state = TelephonyManager.CALL_STATE_IDLE;
            } else if (TelephonyManager.EXTRA_STATE_OFFHOOK.equals(stateStr)) {
                state = TelephonyManager.CALL_STATE_OFFHOOK;
            } else if (TelephonyManager.EXTRA_STATE_RINGING.equals(stateStr)) {
                state = TelephonyManager.CALL_STATE_RINGING;
            }
            onCallStateChanged(context, state, number);
        }
    }

    protected void onIncomingCallReceived(Context context, String number, Date start) {
        context.sendBroadcast(new Intent(INCOMING_CALL_RECEIVED));
    }

    protected void onIncomingCallAnswered(Context context, String number, Date start) {
    }

    protected void onIncomingCallEnded(Context context, String number, Date start, Date end) {
        context.sendBroadcast(new Intent(INCOMING_CALL_ENDED));
    }

    protected void onOutgoingCallStarted(Context context, String number, Date start) {
        context.sendBroadcast(new Intent(OUTGOING_CALL_STARTED));
    }

    protected void onOutgoingCallEnded(Context context, String number, Date start, Date end) {
        context.sendBroadcast(new Intent(OUTGOING_CALL_ENDED));
    }

    protected void onMissedCall(Context context, String number, Date start) {
        context.sendBroadcast(new Intent(INCOMING_CALL_ENDED));
    }

    public void onCallStateChanged(Context context, int state, String number) {
        if (lastState == state) {
            return;
        }
        switch (state) {
            case TelephonyManager.CALL_STATE_RINGING:
                isIncoming = true;
                callStartTime = new Date();
                savedNumber = number;
                onIncomingCallReceived(context, number, callStartTime);
                break;

            case TelephonyManager.CALL_STATE_OFFHOOK:
                if (lastState != TelephonyManager.CALL_STATE_RINGING) {
                    isIncoming = false;
                    callStartTime = new Date();
                    onOutgoingCallStarted(context, savedNumber, callStartTime);
                } else {
                    isIncoming = true;
                    callStartTime = new Date();
                    onIncomingCallAnswered(context, savedNumber, callStartTime);
                }

                break;

            case TelephonyManager.CALL_STATE_IDLE:
                if (lastState == TelephonyManager.CALL_STATE_RINGING) {
                    onMissedCall(context, savedNumber, callStartTime);
                } else if (isIncoming) {
                    onIncomingCallEnded(context, savedNumber, callStartTime, new Date());
                } else {
                    onOutgoingCallEnded(context, savedNumber, callStartTime, new Date());
                }
                break;
        }
        lastState = state;
    }
}

File: questPresentation/src/main/java/ru/nekit/android/qls/lockScreen/receiver/PhoneCallReceiver.kt
package ru.nekit.android.qls.lockScreen.receiver

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.telephony.TelephonyManager.*
import ru.nekit.android.domain.event.IEvent
import ru.nekit.android.domain.event.IEventSender
import ru.nekit.android.qls.QuestLockScreenApplication
import java.util.*

class PhoneCallReceiver : BroadcastReceiver() {

    private fun eventSender(context: Context): IEventSender =
            (context.applicationContext as QuestLockScreenApplication).getEventSender()


    override fun onReceive(context: Context, intent: Intent) {
        val extras = intent.extras!!
        if (Intent.ACTION_NEW_OUTGOING_CALL == intent.action) {
            savedNumber = extras.getString(Intent.EXTRA_PHONE_NUMBER)
        } else {
            val stateString = extras.getString(EXTRA_STATE)
            val number = extras.getString(EXTRA_INCOMING_NUMBER)
            var state = 0
            when (stateString) {
                EXTRA_STATE_IDLE -> state = CALL_STATE_IDLE
                EXTRA_STATE_OFFHOOK -> state = CALL_STATE_OFFHOOK
                EXTRA_STATE_RINGING -> state = CALL_STATE_RINGING
            }
            onCallStateChanged(context, state, number)
        }
    }

    private fun onIncomingCallReceived(context: Context, number: String?, start: Date?) {
        eventSender(context).send(PhoneEvent.INCOMING_CALL_START)
    }

    private fun onIncomingCallAnswered(context: Context, number: String?, start: Date?) {}

    private fun onIncomingCallEnded(context: Context, number: String?, start: Date?, end: Date?) {
        eventSender(context).send(PhoneEvent.INCOMING_CALL_COMPLETE)
    }

    private fun onOutgoingCallStarted(context: Context, number: String?, start: Date?) {
        eventSender(context).send(PhoneEvent.OUTGOING_CALL_START)
    }

    private fun onOutgoingCallEnded(context: Context, number: String?, start: Date?, end: Date?) {
        eventSender(context).send(PhoneEvent.OUTGOING_CALL_COMPLETE)
    }

    private fun onMissedCall(context: Context, number: String?, start: Date?) {
        eventSender(context).send(PhoneEvent.INCOMING_CALL_COMPLETE)
    }

    fun onCallStateChanged(context: Context, state: Int, number: String?) {
        if (lastState == state) {
            return
        }
        when (state) {
            CALL_STATE_RINGING -> {
                isIncoming = true
                callStartTime = Date()
                savedNumber = number
                onIncomingCallReceived(context, number, callStartTime)
            }

            CALL_STATE_OFFHOOK -> if (lastState != CALL_STATE_RINGING) {
                isIncoming = false
                callStartTime = Date()
                onOutgoingCallStarted(context, savedNumber, callStartTime)
            } else {
                isIncoming = true
                callStartTime = Date()
                onIncomingCallAnswered(context, savedNumber, callStartTime)
            }

            CALL_STATE_IDLE -> when {
                lastState == CALL_STATE_RINGING -> onMissedCall(context, savedNumber, callStartTime)
                isIncoming -> onIncomingCallEnded(context, savedNumber, callStartTime, Date())
                else -> onOutgoingCallEnded(context, savedNumber, callStartTime, Date())
            }
        }
        lastState = state
    }

    companion object {


        private var lastState = CALL_STATE_IDLE
        private var callStartTime: Date? = null
        private var isIncoming: Boolean = false
        private var savedNumber: String? = null
    }

    enum class PhoneEvent : IEvent {

        INCOMING_CALL_START,
        INCOMING_CALL_COMPLETE,
        OUTGOING_CALL_START,
        OUTGOING_CALL_COMPLETE;

        override val eventName: String = "${javaClass.name}::$name"

    }
}
